crates/cli/src/bin/wasm-bindgen-test-runner/headless.rs

```diff
@@ -0,0 +1,444 @@
+        if output.len() > 0 {
+            println!("output div contained:\n{}", tab(&output));
+        }
+    }
+    if logs.len() > 0 {
+        println!("console.log div contained:\n{}", tab(&logs));
+    }
+    if errors.len() > 0 {
+        println!("console.log div contained:\n{}", tab(&errors));
+    }
+
+    Ok(())
+}
+
+enum Driver {
+    Gecko(PathBuf),
+    Safari(PathBuf),
+    Chrome(PathBuf),
+}
+
+impl Driver {
+    fn find() -> Result<(Driver, Vec<String>), Error> {
+        let env_args = |name: &str| {
+            env::var(format!("{}_ARGS", name.to_uppercase()))
+                .unwrap_or_default()
+                .split_whitespace()
+                .map(|s| s.to_string())
+                .collect::<Vec<_>>()
+        };
+
+        let drivers = [
+            ("geckodriver", Driver::Gecko as fn(PathBuf) -> Driver),
+            ("safaridriver", Driver::Safari as fn(PathBuf) -> Driver),
+            ("chromedriver", Driver::Chrome as fn(PathBuf) -> Driver),
+        ];
+
+        // First up, if env vars like GECKODRIVER are present, use those to
+        // allow forcing usage of a particular driver.
+        for (driver, ctor) in drivers.iter() {
+            let env = driver.to_uppercase();
+            let path = match env::var_os(&env) {
+                Some(path) => path,
+                None => continue,
+            };
+            return Ok((ctor(path.into()), env_args(driver)))
+        }
+
+        // Next, check PATH. If we can find any supported driver, use that by
+        // default.
+        for path in env::split_paths(&env::var_os("PATH").unwrap_or_default()) {
+            let found = drivers
+                .iter()
+                .find(|(name, _)| {
+                    path.join(name)
+                        .with_extension(env::consts::EXE_EXTENSION)
+                        .exists()
+                });
+            let (name, ctor) = match found {
+                Some(p) => p,
+                None => continue,
+            };
+            return Ok((ctor(name.into()), env_args(name)))
+        }
+
+        // TODO: download an appropriate driver? How to know which one to
+        //       download?
+
+        bail!("failed to find a suitable WebDriver binary to drive headless \
+               testing")
+    }
+
+    fn path(&self) -> &Path {
+        match self {
+            Driver::Gecko(path) => path,
+            Driver::Safari(path) => path,
+            Driver::Chrome(path) => path,
+        }
+    }
+
+    fn browser(&self) -> &str {
+        match self {
+            Driver::Gecko(_) => "Firefox",
+            Driver::Safari(_) => "Safari",
+            Driver::Chrome(_) => "Chrome",
+        }
+    }
+}
+
+struct Client {
+    handle: RefCell<Easy>,
+    driver_addr: SocketAddr,
+}
+
+enum Method<'a> {
+    Get,
+    Post(&'a str),
+    Delete,
+}
+
+impl Client {
+    fn new_session(&self, driver: &Driver) -> Result<String, Error> {
```

Comment:
Did you consider using https://crates.io/crates/fantoccini? I guess that would pull in a lot of dependencies via tokio et al...

Reply:
Yeah I initially started there but it ended up pulling lots of deps we didn't end up really needing, so I opted to go with the hopefully simple route of just handrolling what we need